### DISCUSSION TITLE
eCommerce product variations

### DISCUSSION THREAD

**Discussion by fusionconsulting** [area17/twill#1765] [General]
I'm thinking of using Twill for an upcoming project which will be an eCommerce platform. One of the requirements will be product variations. I have the database structure and everything required for this, I'm wondering if it could work with Twill as the interface. 

I would need to attach one or more options to a product and select from those for example size (Select from S/M/L/XL/XXL etc) and Colour (Select from Blue/Green/Yellow/Orange). Some products may have all options and some only selected ones. Is it possible to make that work using the Twill interface? And, any suggestions on how I might display the variations that are generated from the combinations of the options?

**Comment by kallefrombosnia** [comment]
Hello, yes definitely possible since my team and I made something similar with Twill as an e-commerce admin panel.

Variants were resolved like this.
Every module record (main product) has its own repeater for variants (size variants).
On every repeater, record added we have a browser that gets us results from another module, Colors.

Like this, it works perfectly, but note that we don't have different prices per color. If that's the case, you will need to implement additional logic there to make it work.

I have made a small chart that draws that flow.
![Screenshot_4](https://user-images.githubusercontent.com/30219259/183658510-cede41c6-5ca9-4d9a-b64f-e577d233ddd3.png)

**Reply by fusionconsulting** [reply]
Thanks, that looks promising. How did you setup the attributes? If I wanted to let the user add colours, sizes etc with different options, how could I do that? From my understanding of the above you have modules created in code for the colours and sizes, is that correct? How could I have a "attributes" module and let the user create the attributes to then be selected? Or am I misunderstanding how you have done it?

**Reply by kallefrombosnia** [reply]
It depends on how you want to organize product attributes. We didn't have a task like that, but what I would do is create separate modules for every attribute type (colors, textile type, etc), but that only if attributes are not too dynamic and you have a predefined set of attributes.
And after that in product variants (repeater item) you can include the browser field which will accept multiple modules and save multiple attributes connected to that product variant.

```php
    @formField('browser', [
        'modules' => [
            [
              'label' => 'Colors',
              'name' => 'colors',
            ],
            [
              'label' => 'Textile type',
              'name' => 'textile',
            ],
        ],
        ....
    ])
```

After that, you only need to decide how to parse data and present it on the front end.
Like I said before this will only be good if prices are not connected to the attributes.

For example, a blue XL t-shirt can be more expensive than a red XL t-shirt. If you have problems like this then I would suggest you go with the blocks.
In block, you are free to create any logic wanted and save prices per attribute combination.

**Reply by fusionconsulting** [reply]
Thanks for the detailed response. I think what I need is going to be more complex as users would be able to create their own attributes and options and then create variations based on those with individual stock and pricing. I will have a look at what you have suggested and see if I can make it fit. I think I may end up with a completely custom interface on this though.

**Reply by kallefrombosnia** [reply]
I know this solution can be better, but still, Twill is still CMS, not an ecommerce package.
For this, I can suggest you [https://getcandy.io/ ](https://getcandy.io/) since that is truly e-commerce related package and they are in active development (still not production), but it looks promising.
They also have Discord server, you can ask any question regarding features,
